From ce818f5367f5459c225804b68048d297efdf1f8e Mon Sep 17 00:00:00 2001
From: Ren_Chen <Ren_Chen@wiwynn.com>
Date: Fri, 23 Oct 2020 14:21:53 +0800
Subject: [PATCH] Initialize SCU setting for setting multi-function and GPIO
 pins

Decription:
- Initialize GPIO setting
  Set GPIOA0 and GPIOA1 to GPIO pin
  Set GPIOB5 and GPIOB6 to LPC pin
  Set GPIOR0 to FWSPICS1 ping
  Set GPIOR1 to GPIO pin

Designed:
- Initialize SCU setting for setting multi-function and GPIO pins in ast-g5.c.
---
 board/aspeed/ast-g5/ast-g5.c | 33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/board/aspeed/ast-g5/ast-g5.c b/board/aspeed/ast-g5/ast-g5.c
index 9952d4a38b..b11de688a5 100644
--- a/board/aspeed/ast-g5/ast-g5.c
+++ b/board/aspeed/ast-g5/ast-g5.c
@@ -18,6 +18,14 @@ DECLARE_GLOBAL_DATA_PTR;
 
 #define LPC_REG_BASE              0x1e789000
 #define HOST_INTERFACE_CTL_5      0x80
+#define AST_SCU_BASE              0x1E6E2000
+#define BIT(bit)                  (1 << bit)
+#define MACL1LIN_FUNC_PIN         0
+#define MACL2LIN_FUNC_PIN         1
+#define LPCPD_FUNC_PIN            13
+#define LPCPME_FUNC_PIN           14
+#define FWSPICS1_FUNC_PIN         24
+#define FWSPICS2_FUNC_PIN         25
 
 /* --------------------------------------------------------------------
  * @fn Chip_Register_Init
@@ -46,6 +54,30 @@ void Set_Default_UART_Route(void)
     *(volatile ulong *)(0x1E78909C) = 0x0;
 }
 
+void SCU_Init(void)
+{
+    puts("SCU register Initialization...\n");
+    unsigned long ast_reg32;
+    ast_reg32 = 0x1688A8A8;
+    *(volatile ulong *) (AST_SCU_BASE) = ast_reg32;
+
+    /* Set GPIOA0 and GPIOA1 to GPIO pin
+       Set GPIOB5 and GPIOB6 to LPC pin
+     */
+    ast_reg32 = *(volatile ulong *) (AST_SCU_BASE + 0x80);
+    ast_reg32 |= (BIT(LPCPD_FUNC_PIN)|BIT(LPCPME_FUNC_PIN));
+    ast_reg32 &= ~(BIT(MACL1LIN_FUNC_PIN)|BIT(MACL2LIN_FUNC_PIN));
+    *(volatile ulong *) (AST_SCU_BASE + 0x80) = ast_reg32;
+    /*
+     Set GPIOR0 to FWSPICS1 ping
+     Set GPIOR1 to GPIO pin
+     */
+    ast_reg32 = *(volatile ulong *) (AST_SCU_BASE + 0x88);
+    ast_reg32 |= BIT(FWSPICS1_FUNC_PIN);
+    ast_reg32 &= ~BIT(FWSPICS2_FUNC_PIN);
+    *(volatile ulong *) (AST_SCU_BASE + 0x88) = ast_reg32;
+}
+
 int board_init(void)
 {
 	gd->bd->bi_boot_params = CONFIG_SYS_SDRAM_BASE + 0x100;
@@ -53,6 +85,7 @@ int board_init(void)
 
 	Chip_Register_Init();
 	Set_Default_UART_Route();
+	SCU_Init();
 
 	return 0;
 }
-- 
2.32.0

